### PR TITLE
Fixes #26024,#26025 - removed gem fog & locked versions for fog-core, fog-google, net-ssh

### PR DESCRIPTION
**Pull request by kgaikwad**
There is [`fog-core 2.1.2`]( https://github.com/fog/fog-core) available which contains few deprecation changes related to `Fog::SSH` module which uses net-ssh library but fog version still pointing to 2.1.0 so we might need to update fog version again once they release fog new version.

### CODE REVIEW

**Comment by ares** [comment]
I have verified basic functionality of ec2, libvirt, vmware, ovirt, only the last one will be affected, I spoke with @orrabin who's working on updating fog-ovirt and she's fine with breaking for now to unblock others.

Sadly failures are related, seems we need to update some mock or bump fog-core.

In order to merge this, we need to have tests green. Also @kgaikwad works on PRs for packaging.

**Comment by kgaikwad** [comment]
@ShimShtein,
Updated PR with the change from `expects` to `stubs`.
Please let me know if any additional change required.

File: bundler.d/fog.rb [hunk 1]
```diff
@@ -1,3 +1,3 @@
 group :fog do
-  gem 'fog', '1.42.1', :require => false
+  gem 'fog', '2.1.0', :require => false
```
Comment (ekohl) [inline bundler.d/fog.rb]: Do we actually still need fog itself if we have fog-core and all the various fog-* providers?

**Review by ekohl — COMMENTED** [review]
Should fog-core also be bumped to a newer version?

**Comment by ares** [comment]
Even the newest fog-google will not work with fog-core >= 2.1.0, though fog-aws 3.x requires fog-core >= 2.1.0. I didn't realize until now :-( how about we drop fog-google for now? We know it's broken atm. Then we could bump fog-core, all other resource dependencies seems to be resolvable.

few more findings:
* foreman_docker requires fog, not fog-core, but we can sacrifice foreman_docker I believe after the dependency is removed from Katello (I tested the PR today and it works fine)
* foreman_kubevirt requires older version of fog, that would need to be updated, but it's not part of Foreman org yet

**Comment by kgaikwad** [comment]
@ares, 
For `fog-azure-rm`, it also needs `'fog-core', '~> 1.43'`. Should I remove google changes from this PR?

**Comment by ares** [comment]
So the question is, how confident we are, that we'll be able to make fog-azure-rm and fog-google work with fog-core 2.1.2 (prpbably not fog anymore) before 1.22 branching (7 weeks). If that's concerning, we'd have to revert the fog-aws bump and wait with big fog update. Otoh if we're adding new compute resources, we should start with most recent versions I think, otherwise we're adding technical debt from day 1.

What do you think @kgaikwad?

**Comment by kgaikwad** [comment]
@ares,
I can see `fog-google 1.8.2` is available so I will update this version and will test compute changes.

Related to fog-core 2.1.2, fog-google gives support to fog-core 2.1.0 as "fog-core", "<= 2.1.0" from gemspec so I have to check that how small or big change that would be to update fog-code from 2.1.0 to 2.1.2.
@apuntamb - similarly, could you please check in fog-azure-rm for fog-core version 2.1.2 as it is pointing to older version `~> 1.43`.
After this only we will able to decide whether to go with fog update or wait till other compute resources to bump fog-core to 2.1.2.

**Comment by apuntamb** [comment]
@ares @kgaikwad Hello. 
The latest update for foreman_azure_rm is that I am able to create a host on AzureRM compute resource with the following versions of dependencies required:
**fog = 2.1.0
fog-azure-rm = 0.4.9
fog-core ~> 1.43**
I am currently able to provision a virtual machine (some issues yet to be fixed) but, if I bump `fog-core` to 2.1.0 or later, bundle update fails as fog itself points to `fog-core ~>1.45`. So, as a fix for this, fog can refer to fog-core 2.x, but there are still few dependencies within fog, one for example, fog-joyent which still refers to fog-core as old as `~> 1.42`.
So, if I consider the current scenario, I am pretty confident to make foreman_azure_rm functional as soon as it is expected. And if we are planning to bump fog-core, I guess I should get to testing different dependencies involved.

Please let me know you thoughts.

**Comment by ares** [comment]
@apuntamb so fog-aws requires fog-core 2.1.0, could you retest what you have if you only change dependencies in fog_azure_rm that it depends on fog-core 2.1.0 and no longer on fog? there's a chance it will just work :-)

**Comment by apuntamb** [comment]
@ares Thanks for the guidelines. :-) 
Following are the results that I obtained:
So, if I keep no dependency on fog and simply point my foreman to `fog-core ~> 2.1.2` and fog-azure-rm to `fog-core ~> 2.1.0` ( it still uses fog-core 2.1.2 ), all the functionalities successfully work as earlier. Although if I have `fog-core ~> 2.1.0` in foreman Gemfile, it would fail as `fog-google` has a dependency on `fog-core <= 2.1.0`.

**Comment by ares** [comment]
That's actually good. Now let's see what @kgaikwad finds out about fog-google. If that would be also possible to make it work with fog-core 2.1.2, we could do the upgrade altogether.

**Comment by kgaikwad** [comment]
@ares, 
I tried below scenarios by removing fog gem -
1. With fog-google `1.8.2` and fog-core `2.1.0` which works fine as current gemspec already using this
2. With fog-google `1.8.2` and fog-core `2.1.2` which gives me same deprecation warnings - https://github.com/fog/fog-google/issues/417#issue-357634223

Please let me know if it is okay to lock fog-core to `2.1.0` and once these deprecation messages fixed in code then we can bump fog-core and fog-google again?  

UPDATE - yes, as @apuntamb mentioned I have tested current working steps. Those are working as expected even after deprecation messages.

EDIT - note that with fog removal and katello plugin enabled, it is giving load error for foreman-docker.

**Comment by apuntamb** [comment]
@kgaikwad @ares One similar issue for `fog-core 2.1.2`. I re-ran the server now and received similar deprecation warnings:
```
[fog][DEPRECATION] Unable to load Fog::AzureRM::Resources
11:09:51 rails.1   | [fog][DEPRECATION] The format Fog::Resources::AzureRM is deprecated
11:09:51 rails.1   | [fog][DEPRECATION] Unable to load Fog::AzureRM::DNS
11:09:51 rails.1   | [fog][DEPRECATION] The format Fog::DNS::AzureRM is deprecated
11:09:51 rails.1   | [fog][DEPRECATION] Unable to load Fog::AzureRM::Storage
11:09:51 rails.1   | [fog][DEPRECATION] The format Fog::Storage::AzureRM is deprecated
11:09:51 rails.1   | [fog][DEPRECATION] Unable to load Fog::AzureRM::Network
11:09:51 rails.1   | [fog][DEPRECATION] The format Fog::Network::AzureRM is deprecated
11:09:51 rails.1   | [fog][DEPRECATION] Unable to load Fog::AzureRM::Compute
11:09:51 rails.1   | [fog][DEPRECATION] The format Fog::Compute::AzureRM is deprecated
11:09:51 rails.1   | [fog][DEPRECATION] Unable to load Fog::AzureRM::ApplicationGateway
11:09:51 rails.1   | [fog][DEPRECATION] The format Fog::ApplicationGateway::AzureRM is deprecated
11:09:51 rails.1   | [fog][DEPRECATION] Unable to load Fog::AzureRM::TrafficManager
11:09:51 rails.1   | [fog][DEPRECATION] The format Fog::TrafficManager::AzureRM is deprecated
11:09:51 rails.1   | [fog][DEPRECATION] Unable to load Fog::AzureRM::Sql
11:09:51 rails.1   | [fog][DEPRECATION] The format Fog::Sql::AzureRM is deprecated
11:09:51 rails.1   | [fog][DEPRECATION] Unable to load Fog::AzureRM::KeyVault
11:09:51 rails.1   | [fog][DEPRECATION] The format Fog::KeyVault::AzureRM is deprecated
```
Note - Even with these warnings the functionalities work fine. Also, I have Katello plugin disabled so not receiving any error for foreman-docker.

**Comment by ares** [comment]
yeah, let's ignore foreman-docker for now, we're working on removing it

**Comment by ShimShtein** [comment]
@kgaikwad The tests look good now. No complaints :grinning:

**Comment by ares** [comment]
@kgaikwad I think using fog-core 2.1.0 is fine. but if deprecation messages are just warning, I'd go straight to 2.1.2 so we test the newest version right away

**Comment by kgaikwad** [comment]
@ares,
I have tested `2.1.2` scenario by cloning fog-google repo and modified gemspec file. But in fog-google repo,  gemspec file is still pointing to `2.1.0`. So, pointing fog-core version to  `2.1.0`.

File: app/models/concerns/fog_extensions.rb [hunk 2]
```diff
@@ -15,12 +15,10 @@ module FogExtensions
...
 if Foreman::Model::GCE.available?
   require 'fog/google'
-  require 'fog/google/models/compute/image'
+  require 'fog/compute/google/models/image'
   Fog::Compute::Google::Image.send(:include, FogExtensions::Google::Image)
-  require 'fog/google/models/compute/server'
+  require 'fog/compute/google/models/server'
   Fog::Compute::Google::Server.send(:include, FogExtensions::Google::Server)
-  require 'fog/google/models/compute/flavor'
-  Fog::Compute::Google::Flavor.send(:include, FogExtensions::Google::Flavor)
```
Comment (tbrisker) [inline app/models/concerns/fog_extensions.rb:23]: If this is no longer used should `app/models/concerns/fog_extensions/aws/flavor.rb` be also removed?

**Review by tbrisker — COMMENTED** [review]
minor comment inline, other than that looks good to me - any reason not to merge it yet @kgaikwad @ares ? Is the packaging side ready?

**Comment by ekohl** [comment]
I suspect foreman_docker will need an explicit dependency on fog. There's a good chance without it the app will fail to start up. Given the upgrade scenario will be that the user will upgrade all packages and the try to remove docker (because of the hard dependency in katello), this is something we need to verify.

File: app/models/concerns/fog_extensions.rb (hunk 2 above)
Comment (kgaikwad) [reply app/models/concerns/fog_extensions.rb:23]: As `flavor` model exists in `fog-aws`, we don't need to remove this model.

+1,  If I understood correctly then you would like to mention `app/models/concerns/fog_extensions/google/flavor.rb` here, right?
I will remove the file app/models/concerns/fog_extensions/google/flavor.rb.

**Comment by kgaikwad** [comment]
> minor comment inline, other than that looks good to me - any reason not to merge it yet @kgaikwad @ares ? Is the packaging side ready?

Packaging PR for 'fog-google' is pending on me. Addition to what @ekohl  mentioned, this change is also dependent on katello-side [foreman-docker pr-7677](https://github.com/Katello/katello/pull/7677)

**Comment by lzap** [comment]
I had to add this:

```diff
diff --git a/app/models/concerns/fog_extensions.rb b/app/models/concerns/fog_extensions.rb
index 2ba01a579..1dab2f5cb 100644
--- a/app/models/concerns/fog_extensions.rb
+++ b/app/models/concerns/fog_extensions.rb
@@ -1,6 +1,7 @@
 module FogExtensions
 end
 
+require 'concerns/fog_extensions/model'
 Fog::Model.send(:include, FogExtensions::Model) if defined? Fog::Model
 
 # Fog is required by bundler, and depending on the group configuration,
@@ -41,12 +42,11 @@ end
 
 if Foreman::Model::Openstack.available?
   require 'fog/openstack'
```

In order to prevent this error:

Unable to autoload constant FogExtensions::Model, expected /home/lzap/work/foreman/app/models/concerns/fog_extensions/model.rb to define it (LoadError)

**Comment by kgaikwad** [comment]
@lzap, 
I think its due to openstack, in latest version of `openstack` there is directory change for compute folder.
previously it was pointing to directory `lib/fog/compute/openstack` which is now changed to `lib/fog/openstack/compute`

**Review by ares — APPROVED** [review]
Thanks @kgaikwad, merging!

**Pull request merged**